lib/internal/streams/operators.js

```diff
@@ -215,6 +219,53 @@ async function * filter(fn, options) {
   yield* this.map(filterFn, options);
 }
 
+// Specific to provide better error to reduce since the argument is only
+// missing if the stream has no items in it - but the code is still appropriate
+class ReduceAwareErrMissingArgs extends ERR_MISSING_ARGS {
+  constructor() {
+    super('reduce');
+    this.message = 'Reduce of an empty stream requires an initial value';
+  }
+}
+
+async function reduce(reducer, initialValue, options) {
+  if (typeof reducer !== 'function') {
+    throw new ERR_INVALID_ARG_TYPE(
+      'reducer', ['Function', 'AsyncFunction'], reducer);
+  }
+  let hasInitialValue = arguments.length > 1;
+  // TODO(benjamingr) discuss
+  // The semantics in most of these methods is that the signal being
+  // pre-aborted does not try to consume the stream and thus does not destroy
+  // it - I am not sure that's expected or a good idea.
+  if (options?.signal?.aborted) {
+    throw new AbortError();
+  }
+  const ac = new AbortController();
+  const signal = ac.signal;
+  let gotAnyItemFromStream = false;
+  try {
+    for await (const value of this) {
+      gotAnyItemFromStream = true;
+      if (options?.signal?.aborted) {
+        throw new AbortError();
+      }
+      if (!hasInitialValue) {
+        initialValue = value;
+        hasInitialValue = true;
+      } else {
+        initialValue = await reducer(initialValue, value, { signal });
+      }
+    }
+    if (!gotAnyItemFromStream && !hasInitialValue) {
+      throw new ReduceAwareErrMissingArgs();
```

Comment:
nit
```suggestion
      // Specific to provide better error to reduce since the argument is only
      // missing if the stream has no items in it - but the code is still appropriate
      const error = new ERR_MISSING_ARGS('');
      error.message = 'Reduce of an empty stream requires an initial value';
      throw error;
```
Maybe it's better than having a class that is only used once?

Reply:
Is there a cost to this (having another class)?  I don't mind it either way